// @ably/spaces/dist/cjs/react/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useCursors = exports.useLock = exports.useLocks = exports.useLocations = exports.useMembers = exports.useSpace = exports.useSpaces = exports.SpaceProvider = exports.SpacesProvider = void 0;
var SpacesContext_js_1 = require("./contexts/SpacesContext.js");
Object.defineProperty(exports, "SpacesProvider", { enumerable: true, get: function () { return SpacesContext_js_1.SpacesProvider; } });
var SpaceContext_js_1 = require("./contexts/SpaceContext.js");
Object.defineProperty(exports, "SpaceProvider", { enumerable: true, get: function () { return SpaceContext_js_1.SpaceProvider; } });
var useSpaces_js_1 = require("./useSpaces.js");
Object.defineProperty(exports, "useSpaces", { enumerable: true, get: function () { return useSpaces_js_1.useSpaces; } });
var useSpace_js_1 = require("./useSpace.js");
Object.defineProperty(exports, "useSpace", { enumerable: true, get: function () { return useSpace_js_1.useSpace; } });
var useMembers_js_1 = require("./useMembers.js");
Object.defineProperty(exports, "useMembers", { enumerable: true, get: function () { return useMembers_js_1.useMembers; } });
var useLocations_js_1 = require("./useLocations.js");
Object.defineProperty(exports, "useLocations", { enumerable: true, get: function () { return useLocations_js_1.useLocations; } });
var useLocks_js_1 = require("./useLocks.js");
Object.defineProperty(exports, "useLocks", { enumerable: true, get: function () { return useLocks_js_1.useLocks; } });
var useLock_js_1 = require("./useLock.js");
Object.defineProperty(exports, "useLock", { enumerable: true, get: function () { return useLock_js_1.useLock; } });
var useCursors_js_1 = require("./useCursors.js");
Object.defineProperty(exports, "useCursors", { enumerable: true, get: function () { return useCursors_js_1.useCursors; } });
